File: src/pages/RegistrationPage.ts
import { Page, Locator, expect } from '@playwright/test';
import { LoginPage } from './LoginPage';
import { URLS } from '../config/urls';
import { SELECTORS, TEST_DATA_CONSTANTS, PAGE_TITLES, SUCCESS_MESSAGES, APPLICATION_CONFIG } from '../constants/application-constants';
import * as fs from 'fs';
import * as path from 'path';

/**
 * RegistrationPage class handles user registration functionality in ParaBank
 * 
 * @description This page object class provides methods for user registration,
 * form filling, and credential management. It encapsulates all the locators
 * and business logic related to user registration functionality.
 * 
 * @example
 * ```typescript
 * const registrationPage = new RegistrationPage(page);
 * await registrationPage.registerUser();
 * ```
 */
export class RegistrationPage {
  private page: Page;
  private firstNameField: Locator;
  private lastNameField: Locator;
  private addressField: Locator;
  private cityField: Locator;
  private stateField: Locator;
  private zipCodeField: Locator;
  private phoneNumberField: Locator;
  private ssnField: Locator;
  private usernameField: Locator;
  private passwordField: Locator;
  private confirmPasswordField: Locator;
  private registerButton: Locator;
  private logout: Locator;
  private errorMessageLocator: Locator;

  /**
   * Creates an instance of RegistrationPage
   * 
   * @param {Page} page - Playwright page instance for browser interactions
   * 
   * @description Initializes the page object with all necessary locators for
   * user registration form fields and buttons.
   */
  constructor(page: Page) {
    this.page = page;
    this.firstNameField = page.locator(SELECTORS.REGISTRATION.FIRST_NAME);
    this.lastNameField = page.locator(SELECTORS.REGISTRATION.LAST_NAME);
    this.addressField = page.locator(SELECTORS.REGISTRATION.ADDRESS);
    this.cityField = page.locator(SELECTORS.REGISTRATION.CITY);
    this.stateField = page.locator(SELECTORS.REGISTRATION.STATE);
    this.zipCodeField = page.locator(SELECTORS.REGISTRATION.ZIP_CODE);
    this.phoneNumberField = page.locator(SELECTORS.REGISTRATION.PHONE);
    this.ssnField = page.locator(SELECTORS.REGISTRATION.SSN);
    this.usernameField = page.locator(SELECTORS.REGISTRATION.USERNAME);
    this.passwordField = page.locator(SELECTORS.REGISTRATION.PASSWORD);
    this.confirmPasswordField = page.locator(SELECTORS.REGISTRATION.CONFIRM_PASSWORD);
    this.registerButton = page.locator(SELECTORS.REGISTRATION.REGISTER_BUTTON);
    this.logout = page.locator(SELECTORS.REGISTRATION.LOGOUT_LINK);
    this.errorMessageLocator = page.locator(".error");
  }

  /**
   * Generates a random username for user registration
   * 
   * @description This method creates a unique username by combining a random word
   * from a predefined list with a random 4-digit number and appending @gmail.com.
   * This ensures uniqueness for each test execution.
   * 
   * @returns {string} A randomly generated username in email format
   * 
   * @example
   * ```typescript
   * const username = this.generateRandomUsername();
   * // Returns: "alpha1234@gmail.com"
   * ```
   */
  private generateRandomUsername(): string {
    const words = ["alpha", "beta", "gamma", "delta", "omega", "zeta"];
    const numbers = Math.floor(Math.random() * 9000) + 1000;
    return `${words[Math.floor(Math.random() * words.length)]}${numbers}@gmail.com`;
  }

  /**
   * Registers a new user with random credentials and saves them to file
   * 
   * @description This method performs a complete user registration process including:
   * - Generating random username and password
   * - Filling all registration form fields with test data
   * - Submitting the registration form
   * - Handling loading pages and success verification
   * - Saving credentials to JSON file for later use
   * - Logging out and logging back in to verify registration
   * 
   * @returns {Promise<void>} Promise that resolves when registration is complete
   * 
   * @throws {Error} Throws an error if registration fails or success indicators not found
   * 
   * @example
   * ```typescript
   * await registrationPage.registerUser();
   * // User is now registered and logged in
   * ```
   */
  async registerUser(): Promise<void> {
    const username = this.generateRandomUsername();
    const password = TEST_DATA_CONSTANTS.DEFAULT_PASSWORD;

    await this.firstNameField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.FIRST_NAME);
    await this.lastNameField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.LAST_NAME);
    await this.addressField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.ADDRESS);
    await this.cityField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.CITY);
    await this.stateField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.STATE);
    await this.zipCodeField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.ZIP_CODE);
    await this.phoneNumberField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.PHONE_NUMBER);
    await this.ssnField.fill(TEST_DATA_CONSTANTS.TEST_PROFILE.SSN);
    await this.usernameField.fill(username);
    await this.passwordField.fill(password);
    await this.confirmPasswordField.fill(password);
    await this.registerButton.click();
    await this.page.waitForLoadState('networkidle', { timeout: APPLICATION_CONFIG.TIMEOUTS.DEFAULT }); // Allow time for registration

    const baseDir = path.dirname(path.dirname(__dirname));
    const dataDir = path.join(baseDir, "data");
    const filePath = path.join(dataDir, "credentials.json");

    // Ensure directory exists
    if (!fs.existsSync(dataDir)) {
      fs.mkdirSync(dataDir, { recursive: true });
    }

    const credentials = { username, password };
    fs.writeFileSync(filePath, JSON.stringify(credentials, null, 2));

    // Wait for the page to load and check for success indicators
    await this.page.waitForLoadState('domcontentloaded');
    
    // Check if we're on a loading page first
    const currentTitle = await this.page.title();
    console.log("Current page title:", currentTitle);
    
    if (currentTitle.includes("Just a moment")) {
      console.log("Detected loading page, waiting for redirect...");
      await this.page.waitForLoadState('networkidle'); // Wait longer for redirect
    }
    
        // Try to find success message or check for accounts overview
        const successMessage = this.page.locator(`text=${SUCCESS_MESSAGES.REGISTRATION_SUCCESS}`);
        const accountsOverview = this.page.locator("text=Accounts Overview");
    
    try {
      await expect(successMessage).toBeVisible({ timeout: 10000 });
      console.log("Registration success message found!");
    } catch {
      try {
        await expect(accountsOverview).toBeVisible({ timeout: 5000 });
        console.log("Accounts overview found - registration successful!");
      } catch {
        console.log("Registration may have succeeded but success indicators not found");
      }
    }

    await this.logout.click();
    await this.page.goto(URLS.HOME);
    const loginPage = new LoginPage(this.page);
    await this.page.waitForLoadState('networkidle');
    await loginPage.login(username, password);
        await expect(this.page).toHaveTitle(PAGE_TITLES.ACCOUNTS_OVERVIEW);
    console.log("Login successful!");

    console.log(`✅ Registration successful! Credentials saved in credentials.json`);
  }
}

File: src/pages/registration_page.py
"""
Registration page object for ParaBank application.
"""
import os
import json
from playwright.sync_api import Page, expect
from src.pages.base_page import BasePage
from src.pages.login_page import LoginPage
from src.config.urls import URLs
from src.constants.application_constants import (
    SELECTORS, TEST_DATA_CONSTANTS, PAGE_TITLES, 
    SUCCESS_MESSAGES, APPLICATION_CONFIG
)


class RegistrationPage(BasePage):
    """Registration page class for ParaBank user registration functionality."""
    
    def __init__(self, page: Page):
        """
        Initialize the registration page.
        
        Args:
            page: Playwright page instance
        """
        super().__init__(page)
        self.first_name_field = page.locator(SELECTORS['REGISTRATION']['FIRST_NAME'])
        self.last_name_field = page.locator(SELECTORS['REGISTRATION']['LAST_NAME'])
        self.address_field = page.locator(SELECTORS['REGISTRATION']['ADDRESS'])
        self.city_field = page.locator(SELECTORS['REGISTRATION']['CITY'])
        self.state_field = page.locator(SELECTORS['REGISTRATION']['STATE'])
        self.zip_code_field = page.locator(SELECTORS['REGISTRATION']['ZIP_CODE'])
        self.phone_number_field = page.locator(SELECTORS['REGISTRATION']['PHONE'])
        self.ssn_field = page.locator(SELECTORS['REGISTRATION']['SSN'])
        self.username_field = page.locator(SELECTORS['REGISTRATION']['USERNAME'])
        self.password_field = page.locator(SELECTORS['REGISTRATION']['PASSWORD'])
        self.confirm_password_field = page.locator(SELECTORS['REGISTRATION']['CONFIRM_PASSWORD'])
        self.register_button = page.locator(SELECTORS['REGISTRATION']['REGISTER_BUTTON'])
        self.logout = page.locator(SELECTORS['REGISTRATION']['LOGOUT_LINK'])
        self.error_message_locator = page.locator(".error")
    
    def _generate_random_username(self) -> str:
        """
        Generate a random username for user registration.
        
        Returns:
            A randomly generated username in email format
        """
        words = ["alpha", "beta", "gamma", "delta", "omega", "zeta"]
        numbers = str(1000 + (hash(str(self.page)) % 9000))  # Use page hash for uniqueness
        return f"{words[hash(str(self.page)) % len(words)]}{numbers}@gmail.com"
    
    def register_user(self) -> None:
        """
        Register a new user with random credentials and save them to file.
        
        This method performs a complete user registration process including:
        - Generating random username and password
        - Filling all registration form fields with test data
        - Submitting the registration form
        - Handling loading pages and success verification
        - Saving credentials to JSON file for later use
        - Logging out and logging back in to verify registration
        """
        username = self._generate_random_username()
        password = TEST_DATA_CONSTANTS['DEFAULT_PASSWORD']
        
        # Fill registration form
        self.first_name_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['FIRST_NAME'])
        self.last_name_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['LAST_NAME'])
        self.address_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['ADDRESS'])
        self.city_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['CITY'])
        self.state_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['STATE'])
        self.zip_code_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['ZIP_CODE'])
        self.phone_number_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['PHONE_NUMBER'])
        self.ssn_field.fill(TEST_DATA_CONSTANTS['TEST_PROFILE']['SSN'])
        self.username_field.fill(username)
        self.password_field.fill(password)
        self.confirm_password_field.fill(password)
        
        self.register_button.click()
        self.page.wait_for_load_state('networkidle', timeout=APPLICATION_CONFIG['TIMEOUTS']['DEFAULT'])
        
        # Save credentials to file
        self._save_credentials(username, password)
        
        # Wait for the page to load and check for success indicators
        self.page.wait_for_load_state('domcontentloaded')
        
        # Check if we're on a loading page first
        current_title = self.page.title()
        print(f"Current page title: {current_title}")
        
        if "Just a moment" in current_title:
            print("Detected loading page, waiting for redirect...")
            self.page.wait_for_load_state('networkidle')
        
        # Try to find success message or check for accounts overview
        success_message = self.page.locator(f"text={SUCCESS_MESSAGES['REGISTRATION_SUCCESS']}")
        accounts_overview = self.page.locator("text=Accounts Overview")
        
        try:
            expect(success_message).to_be_visible(timeout=10000)
            print("Registration success message found!")
        except:
            try:
                expect(accounts_overview).to_be_visible(timeout=5000)
                print("Accounts overview found - registration successful!")
            except:
                print("Registration may have succeeded but success indicators not found")
        
        # Logout and login again to verify registration
        self.logout.click()
        self.navigate_to(URLs.HOME)
        login_page = LoginPage(self.page)
        self.page.wait_for_load_state('networkidle')
        login_page.login(username, password)
        expect(self.page).to_have_title(PAGE_TITLES['ACCOUNTS_OVERVIEW'])
        print("Login successful!")
        
        print(f"✅ Registration successful! Credentials saved in credentials.json")
    
    def _save_credentials(self, username: str, password: str) -> None:
        """
        Save user credentials to JSON file.
        
        Args:
            username: Username to save
            password: Password to save
        """
        base_dir = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
        data_dir = os.path.join(base_dir, "data")
        file_path = os.path.join(data_dir, "credentials.json")
        
        # Ensure directory exists
        os.makedirs(data_dir, exist_ok=True)
        
        credentials = {"username": username, "password": password}
        with open(file_path, 'w') as f:
            json.dump(credentials, f, indent=2)
